### PR TITLE
Actionhero v28

### PR DESCRIPTION
**Pull request by evantahler**
### Development

- [x] Application changes have been tested appropriately

### Impact

- [x] Code follows company security practices and guidelines
- [x] Security impact of change has been considered
- [x] Performance impact of change has been considered
- [x] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

### CODE REVIEW

File: core/__tests__/modules/codeConfig/codeConfig.ts [hunk 1]
```diff
@@ -711,9 +712,12 @@ describe("modules/codeConfig", () => {
...
 
-      previousGroupRun = await specHelper.runTask("group:destroy", {
-        groupId: "high_value",
-      });
+      previousGroupRun = await specHelper.runTask<GroupDestroy>(
+        "group:destroy",
+        {
+          groupId: "high_value",
+        }
+      );
```
Comment (evantahler) [inline core/__tests__/modules/codeConfig/codeConfig.ts:720]: `specHelper.runTask` has return types now!

File: core/src/index.ts [hunk 2]
```diff
@@ -58,3 +59,20 @@ export { waitForLock } from "./modules/locks";
...
+import "./config/next";
+import "./config/pluginInjection";
+import "./config/plugins";
+import "./config/redis";
+import "./config/routes";
+import "./config/sequelize";
+import "./config/tasks";
+import "./config/telemetry";
+import "./config/web";
+import "./config/websocket";
```
Comment (evantahler) [inline core/src/index.ts:78]: This was an interesting trick - to force consumers (aka plugins) to be aware of the config types, they need to source the module overrides in the config files.  Adding them here in the main export file makes that happen!

File: core/src/actions/swagger.ts [hunk 3]
```diff
@@ -54,76 +49,91 @@ export class Swagger extends Action {
...
+        if (!action) return;
+
+        const tag = action.name.split(":")[0];
+        const formattedPath = route.path
+          .replace("/v:apiVersion", "")
+          .replace(/\/:(\w*)/, "/{$1}")
+          .replace(/\/:(\w*)/, "/{$1}")
+          .replace(/\/:(\w*)/, "/{$1}")
+          .replace(/\/:(\w*)/, "/{$1}")
+          .replace(/\/:(\w*)/, "/{$1}");
```
Comment (pedroslopez) [inline core/src/actions/swagger.ts]:
```suggestion
          .replace(/\/:(\w*)/g, "/{$1}")
```

Use global flag to replace all?

File: core/src/config/api.ts [hunk 4]
```diff
@@ -105,24 +113,21 @@ export const DEFAULT = {
...
     };
   },
 };
 
 export const production = {
-  general: (config) => {
+  [namespace]: () => {
     return {
       fileRequestLogLevel: "debug",
-      developmentMode: false,
```
Comment (pedroslopez) [inline core/src/config/api.ts:125]: No idea what this is for, but noting that it's been removed from `production`

File: core/src/config/errors.ts [hunk 5]
```diff
@@ -16,48 +26,57 @@ export const DEFAULT = {
...
 
       // action not useable by this client/server type
-      unsupportedServerType: (data) => {
+      unsupportedServerType: (data: ActionProcessor<any>) => {
         return `this action does not support the ${data.connection.type} connection type`;
       },
 
       // action failed because server is mid-shutdown
-      serverShuttingDown: (data) => {
+      serverShuttingDown: (data: ActionProcessor<any>) => {
```
Comment (pedroslopez) [inline core/src/config/errors.ts]: [suggestion] these unused params can probably be removed

File: core/src/config/logger.ts [hunk 6]
```diff
@@ -1,15 +1,34 @@
...
 
 export const DEFAULT = {
-  logger: (config) => {
-    const loggers = [];
-    loggers.push(buildConsoleLogger(process.env.GROUPAROO_LOG_LEVEL));
-    config.general.paths.log.forEach((p) => {
-      loggers.push(buildFileLogger(p));
+  [namespace]: (config: ActionheroConfigInterface) => {
+    const loggers: ActionheroConfigLoggerBuilderArray = [];
+    loggers.push(buildConsoleLogger(process.env.LOG_LEVEL));
```
Comment (pedroslopez) [inline core/src/config/logger.ts]: Not sure if this was intentional, but this is changing the log level env var from `GROUPAROO_LOG_LEVEL` to `LOG_LEVEL`, which may have an impact in other places. I would suggest keeping it as `GROUPAROO_LOG_LEVEL` if possible.

File: plugins/@grouparoo/demo/src/util/runAction.ts [hunk 7]
```diff
@@ -1,6 +1,3 @@
-//import { Connection, Action, config, log, utils, api, action } from 'actionhero';
-//import * as dotProp from "dot-prop";
```
Comment (pedroslopez) [inline plugins/@grouparoo/demo/src/util/runAction.ts:2]: it looks like this `dotProp` thing was being used in the commented out code below - should all that commented code be removed too?

**Review by pedroslopez — CHANGES_REQUESTED** [review]
Nice! I think the main thing to look at is the `LOG_LEVEL` environment var change that may potentially affect users and is referenced in multiple other places in the app.

File: core/src/config/api.ts (hunk 4 above)
Comment (evantahler) [reply core/src/config/api.ts:125]: I think that was a mistake - we don't want it (it's how the hot-reloading works)

File: core/src/config/errors.ts (hunk 5 above)
Comment (evantahler) [reply core/src/config/errors.ts]: Yeah - this one was tricky! In the templates Actionhero generates, I went with generating these methods with all possible args... so now we have them :/

Comment (evantahler) [reply core/src/config/errors.ts]: I'll remove

File: plugins/@grouparoo/demo/src/util/runAction.ts (hunk 7 above)
Comment (evantahler) [reply plugins/@grouparoo/demo/src/util/runAction.ts:2]: I think I'll leave it for now... not really sure what @bleonard was up to here :D

File: core/src/actions/swagger.ts (hunk 3 above)
Comment (evantahler) [reply core/src/actions/swagger.ts]: I *think* this was a problem with the `$1` replacement value - we don't know which index to use to replace things :/   
There's probably a loop we can use here to get them *all* - want to give it a try in the Actionhero repo? 

(edit) made an issue in the Actionhero repo - https://github.com/actionhero/actionhero/issues/2038

File: core/src/config/logger.ts (hunk 6 above)
Comment (evantahler) [reply core/src/config/logger.ts]: Good catch!

**Comment by evantahler** [comment]
@pedroslopez proved that https://github.com/grouparoo/grouparoo/pull/2635/commits/02cf895bcd8ef89776a89f3e09b0c42bc52fb949 was a good idea via https://github.com/actionhero/actionhero/pull/2039

File: core/src/actions/swagger.ts [hunk 8]
```diff
@@ -82,13 +82,7 @@ export class Swagger extends Action {
         if (!action) return;
 
         const tag = action.name.split(":")[0];
-        const formattedPath = route.path
-          .replace("/v:apiVersion", "")
```
Comment (pedroslopez) [inline core/src/actions/swagger.ts]: A bit too eager on the deletion there, I believe`.replace("/v:apiVersion", "")` should still be present 😛

**Review by pedroslopez — CHANGES_REQUESTED** [review]

File: core/src/actions/swagger.ts (hunk 8 above)
Comment (evantahler) [reply core/src/actions/swagger.ts]: oops! updated now

**Review by pedroslopez — APPROVED** [review]

**Pull request merged**